`package_base`: generify accessor methods for `when`-keyed dictionaries

tgamblin opened the pull request:
This turns some variant-specific methods for dealing with when-keyed dictionaries into more generic versions, in preparation for conditional version definitions.

`_by_name`, `_names`, etc. are replaced with generic methods for transforming when-keyed dictionaries:
 * `_by_subkey()`
 * `_subkeys()`
 * `_num_definitions()`
 * `_definitions()`
 * `_remove_overridden_defs()`

And the variant accessors are refactored to use these methods underneath.

To do this, types like `WhenDict` had to be generified, and some `TypeVars` were added for sortable keys and values.

There are 2 commits here that should be rebased -- one vendors `typing_extensions`, the other makes this change.  `typing_extensions` should be useful for us in other places, as it provides `Literal`, `TypedDict`, `Protocol`, and other useful classes.

tgamblin committed:
`package_base`: generify accessor methods for `when`-keyed dictionaries

This turns some variant-specific methods for dealing with when-keyed dictionaries into
more generic versions, in preparation for conditional version definitions.

`_by_name`, `_names`, etc. are replaced with generic methods for transforming
when-keyed dictionaries:
 * `_by_subkey()`
 * `_subkeys()`
 * `_num_definitions()`
 * `_definitions()`
 * `_remove_overridden_defs()`

And the variant accessors are refactored to use these methods underneath.

To do this, types like `WhenDict` had to be generified, and some `TypeVars`
were added for sortable keys and values.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/spack/package_base.py:
```diff
@@ -25,7 +25,23 @@
+    Union,
+)
+
+if TYPE_CHECKING:
+    from typeshed import SupportsRichComparisonT
```
Where do we get the `typeshed` import from?

tgamblin replied:
It’s part of `mypy` and has to be there if you’re running `mypy`, so I think it’s escape to use it conditionally like this.

See https://github.com/python/mypy/tree/master/mypy

alalazo replied:
Can we use something based on `typing.Protocol` instead? That would make us independent of `mypy` in the annotations, and we could avoid conditional imports.

alalazo replied:
Also, I think using the above, or even my suggestion, makes style checks possible only for Python 3.8 and above. I'll double check that.

tgamblin replied:
`typing.Protocol` is also only in 3.8, which is why I suspect `SupportsRichComparisonT` doesn't exist until then -- it's a protocol.  I'm not sure how we should do this other than vendor `typing_extensions`, which is maybe a good idea anyway.  It would get us typed dicts, literals and other stuff I've wanted in the past.

Ok, I vendored `typing_extensions` and can use `Protocol` -- that was easy and it's only one file. Of course the annoying part is that even in newer pythons, protocols like `SupportsRichComparisonT` isn't defined outside `typeshed`, and googling this tells me that I'm not supposed to import directly from `typeshed`... so I have to make my own version of the protocol even though it's the thing that `mypy` uses to describe types (I found it because this was the type bound it was complaining about).

What's the typical way to deal with this?  Do people just make their own protocols?  Maybe @adamjstewart knows.

Will work on making my own protocol...

tgamblin replied:
Ok I implemented this with a protocol -- see what you think.

alalazo reviewed (commented):
LGTM, only one question on `typeshed`

tgamblin committed:
Vendor `typing_extensions`

We are using more and more typing features in Spack, and without features like
protocols, typing core is becoming harder and harder.

I think it's worth vendoring `typing_extensions` for this. It will get us a number of
useful capabilities:

* `Literal`
* `TypedDict`
* `Protocol`

among others.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/spack/util/typing.py:
```diff
@@ -0,0 +1,24 @@
+class SupportsRichComparison(Protocol):
+    """Objects that support < and >"""
+
+    def __eq__(self, other: Any) -> bool: ...
+
+    def __ne__(self, other: Any) -> bool: ...
+
+    def __lt__(self, other: Any) -> bool: ...
+
+    def __le__(self, other: Any) -> bool: ...
+
+    def __gt__(self, other: Any) -> bool: ...
+
+    def __ge__(self, other: Any) -> bool: ...
```
This is minor, but these methods should raise `NotImplementedError`, see https://typing.readthedocs.io/en/latest/spec/protocol.html#protocol-members

tgamblin replied:
done

alalazo commented on lib/spack/external/vendor.txt:
```diff
@@ -8,3 +8,4 @@ six==1.16.0
 macholib==1.16.2
   altgraph==0.17.3
 ruamel.yaml==0.17.21
+typing_extensions==4.1.1
```
Any specific reason why we are using 4.1.1 and not 4.2.0 (just curious)?

tgamblin replied:
4.1.1 is the last version to support 3.6, 4.2.0 says it requires >= 3.7
* https://pypi.org/project/typing-extensions/4.1.1/
* https://pypi.org/project/typing-extensions/4.2.0/

alalazo replied:
Oh, you're right, sorry for the noise:

https://github.com/python/typing/blob/9a39406fa45fbfa76eb5d35afb7863f5f727ad87/typing_extensions/pyproject.toml#L12

I was confused by this:

https://github.com/python/typing/blob/9a39406fa45fbfa76eb5d35afb7863f5f727ad87/typing_extensions/pyproject.toml#L38

but apparently it's just a leftover
